### PR TITLE
Homework1 

### PR DESCRIPTION
**Pull request by AlderaminCph**

### CODE REVIEW

**Comment by AlderaminCph** [comment]
All tasks from homework1 was included in pull request

File: homework1/task02.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+from typing import Sequence
+
+
+def check_fibonacci(data: Sequence[int]) -> bool:
+    """
+    function that checks if the given sequence is a Fibonacci sequence
+    returns Bool
+    """
+    if len(data) >= 3 and data[1] != 0:
```
Comment (annkupriyanova) [inline homework1/task02.py:23]: I would check both the first and the second elements  of a sequence

File: tests/homework1/test_task03.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import pytest
+
+from homework1.task03 import find_maximum_and_minimum
+
+
+@pytest.mark.parametrize(
+    ["input_file", "expected_result"],
+    [
+        (os.path.join("tests/homework1/input_files", "input1.txt"), (0, 12)),
```
Comment (annkupriyanova) [inline tests/homework1/test_task03.py]: I would set a variable for files directory: "tests/homework1/input_files"

File: homework1/task05.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+def find_maximal_subarray_sum(nums: List[int], k: int) -> int:
+    """
+    function that finds largest subarray sum
+    with at-least k elements in it.
+    Returns maximum sum of a subarray
+    with at-least k elements.
+    """
+    if len(nums) == 0 or k <= 0 or k > len(nums):
+        return "Invalid input: k should be nonzero and less than nonzero array size."
```
Comment (annkupriyanova) [inline homework1/task05.py]: I would raise an Exception in this case

File: homework1/sample_project/calculator/calc.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    4 create a homework1 directory in the repo
+    5 then copy the sample_project into the directory.
+    6 fix all bugs in the sample_project code
+    7 write an extra test for each found bug
+"""
+
+
+def check_power_of_2(a: int) -> bool:
+    """
+    function that checks if the number is the power of 2
```
Comment (annkupriyanova) [inline homework1/sample_project/calculator/calc.py]: Use [PEP257 ](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings)in docstrings formatting

**Review by annkupriyanova — CHANGES_REQUESTED** [review]

File: tests/homework1/test_task03.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import os
+from typing import Tuple
+
+import pytest
+
+from homework1.task03 import find_maximum_and_minimum
+
+path = "tests/homework1/input_files"
```
Comment (annkupriyanova) [inline tests/homework1/test_task03.py]: It's better to call the variable more specific not to collapse with os.path name. For example, `test_files_path`

File: homework1/task05.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+def find_maximal_subarray_sum(nums: List[int], k: int) -> int:
+    """
+    function that finds largest subarray sum
+    with at-least k elements in it.
+    Returns maximum sum of a subarray
+    with at-least k elements.
+    """
+    if len(nums) == 0 or k <= 0 or k > len(nums):
+        raise ValueError
```
Comment (annkupriyanova) [inline homework1/task05.py]: The good rule is to add a custom message to an exception

File: tests/homework1/test_task05.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import pytest
+
+from homework1.task05 import find_maximal_subarray_sum
```
Comment (annkupriyanova) [inline tests/homework1/test_task05.py:3]: Let's also add a test for exception path in the function

**Review by annkupriyanova — CHANGES_REQUESTED** [review]

File: homework1/sample_project/calculator/calc.py (hunk 4 above)
Comment (annkupriyanova) [reply homework1/sample_project/calculator/calc.py]: Use  it in all functions

File: homework1/task05.py [hunk 8]
```diff
@@ -20,7 +20,7 @@ def find_maximal_subarray_sum(nums: List[int], k: int) -> int:
         maximum sum of a subarray with at-least k elements.
     """
     if len(nums) == 0 or k <= 0 or k > len(nums):
-        raise ValueError
+        raise ValueError("k should not be greater then the size of list of integers")
```
Comment (annkupriyanova) [inline homework1/task05.py]: When you have multiple conditions, it's more concise to say what values `k` can take and print the actual value of `k`. Otherwise, you need to mention all of the conditions.

File: tests/homework1/test_task05.py [hunk 9]
```diff
@@ -15,3 +15,7 @@ def test_find_maximal_subarray_sum(nums: list, k: int, expected_result: int):
     our_result = find_maximal_subarray_sum(nums, k)
 
     assert our_result == expected_result
+
+def test_find_maximal_subarray_sum_exception():
```
Comment (annkupriyanova) [inline tests/homework1/test_task05.py:20]: Also add test cases with negative `k` and empty array to check all of the conditions.

**Review by annkupriyanova — CHANGES_REQUESTED** [review]

File: homework1/task05.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+The written function should return the sum of this sub-array.
+Examples:
+    nums = [1, 3, -1, -3, 5, 3, 6, 7], k = 3
+    result = 16
+"""
+from typing import List
+
+
+def find_maximal_subarray_sum(nums: List[int], k: int) -> int:
+    """Find largest subarray sum with at-least k elements in it.
```
Comment (annkupriyanova) [inline homework1/task05.py]: Why "at least"? In the task it's said "less [or] equal to k"

File: homework1/task05.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    Args:
+        nums: list of integers.
+        k: number of elements of subarray.
+    Returns:
+        maximum sum of a subarray with at-least k elements.
+    """
+    if len(nums) == 0 or k <= 0 or k > len(nums):
+        raise ValueError("k should not be greater then the size of list of integers")
+    summ = 0
+    for i in range(k):
```
Comment (annkupriyanova) [inline homework1/task05.py]: Do you know how to replace the lines 24-26 with just one line using `sum()` function and list slicing?

File: homework1/task05.py (hunk 10 above)
Comment (annkupriyanova) [reply homework1/task05.py]: For example, if we have nums = [1, 3, -1, -3, 5, -3, 6, 7] and k=3, I guess the result should be 13 (6+7)

**Review by annkupriyanova — CHANGES_REQUESTED** [review]

**Review by annkupriyanova — APPROVED** [review]
